src/hotspot/share/prims/universalUpcallHandler.cpp

```diff
@@ -25,6 +25,66 @@
 #include "prims/universalUpcallHandler.hpp"
 #include "runtime/jniHandles.inline.hpp"
 #include "runtime/interfaceSupport.inline.hpp"
+#include "runtime/javaCalls.hpp"
+#include "classfile/symbolTable.hpp"
+
+#define FOREIGN_ABI "jdk/internal/foreign/abi/"
+
+extern struct JavaVM_ main_vm;
+
+JNI_ENTRY(void, ProgrammableUpcallHandler::upcall_helper(JNIEnv* env, jobject rec, address buff))
+  const UpcallMethod& upcall_method = instance().upcall_method;
+
+  ResourceMark rm(thread);
+  JavaValue result(T_VOID);
+  JavaCallArguments args(2); // long = 2 slots
+
+  args.push_jobject(rec);
+  args.push_long((jlong) buff);
+
+  JavaCalls::call_static(&result, upcall_method.klass, upcall_method.name, upcall_method.sig, &args, thread);
+JNI_END
+
+void ProgrammableUpcallHandler::attach_thread_and_do_upcall(jobject rec, address buff) {
+  Thread* thread = Thread::current_or_null();
+  bool should_detach = false;
+  JNIEnv* p_env = nullptr;
+  if (thread == nullptr) {
+    JavaVM_ *vm = (JavaVM *)(&main_vm);
+    vm->functions->AttachCurrentThread(vm, (void**) &p_env, nullptr);
```

Comment:
The return value should be checked in case of errors.
Making this non-daemon now is probably somewhat safer.